File: app/src/main/java/com/jacky/beedee/ui/widget/RowItemView.java
package com.jacky.beedee.ui.widget;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.jacky.beedee.R;

/**
 * 2018/11/2.
 * GitHub:[https://github.com/jacky1234]
 *
 * @author jacky
 */
public class RowItemView extends RelativeLayout {
    public final static int FLAG_NONE = 0;      //仅仅左边的title
    public final static int FLAG_RIGHT_SHOW_IMAGE = 1;
    public final static int FLAG_RIGHT_SHOW_TEXT = 1 << 1;
    public final static int FLAG_RIGHT_EDITABLE = 1 << 2;
    public final static int FLAG_RIGHT_CHECKABLE = 1 << 3;

    private int flag = FLAG_RIGHT_SHOW_TEXT;

    private TextView tv_title;
    private View iv_arrow;
    private ImageView iv_image;
    private TextView tv_right_desc;
    private EditText et_edit;
    private CheckBox checkBox;

    public RowItemView(Context context) {
        this(context, null);
    }

    public RowItemView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RowItemView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        LayoutInflater.from(context).inflate(R.layout.row_list_item, this, true);

        tv_title = findViewById(R.id.tv_title);
        iv_arrow = findViewById(R.id.iv_arrow);
        iv_image = findViewById(R.id.tv_image);
        iv_image = findViewById(R.id.tv_image);
        et_edit = findViewById(R.id.et_edit);
        tv_right_desc = findViewById(R.id.tv_right_desc);
        checkBox = findViewById(R.id.checkBox);
    }

    public void setType(int type) {
        tv_right_desc.setVisibility(View.GONE);
        iv_image.setVisibility(GONE);
        et_edit.setVisibility(View.GONE);
        checkBox.setVisibility(View.GONE);

        this.flag = type;
        iv_image.setVisibility((flag & FLAG_RIGHT_SHOW_IMAGE) != 0 ? VISIBLE : GONE);
        tv_right_desc.setVisibility((flag & FLAG_RIGHT_SHOW_TEXT) != 0 ? VISIBLE : GONE);
        if ((flag & FLAG_RIGHT_EDITABLE) != 0) {
            iv_arrow.setVisibility(View.GONE);
            et_edit.setVisibility(View.VISIBLE);
        }

        if ((flag & FLAG_RIGHT_CHECKABLE) != 0) {
            iv_arrow.setVisibility(View.GONE);
            checkBox.setVisibility(View.VISIBLE);
        }
    }

    public void setTitle(String text) {
        tv_title.setText(text);
    }

    public void setRightDesc(String text) {
        tv_right_desc.setText(text);
    }

    public ImageView getImageView() {
        return iv_image;
    }

    public void setRightEditableText(CharSequence text) {
        et_edit.setText(text);
    }

    public CharSequence getRightEditableContent() {
        return et_edit.getText();
    }

    public EditText getRightEditText() {
        return et_edit;
    }

    public void setCheckedChangeListener(@Nullable CompoundButton.OnCheckedChangeListener listener) {
        checkBox.setOnCheckedChangeListener(listener);
    }

    public void setChecked(boolean flag) {
        checkBox.setChecked(flag);
    }
}

File: app/src/main/java/com/jacky/labeauty/ui/widget/RowItemView.kt
package com.jacky.labeauty.ui.widget

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import com.jacky.labeauty.R

/**
 * 2018/11/2.
 * GitHub:[https://github.com/jacky1234]
 *
 * @author jacky
 */
class RowItemView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : RelativeLayout(context, attrs, defStyleAttr) {

    private var flag = FLAG_RIGHT_SHOW_TEXT

    private val tv_title: TextView
    private val iv_arrow: View
    var imageView: ImageView? = null
        private set
    private val tv_right_desc: TextView
    val rightEditText: EditText
    private val checkBox: CheckBox

    val rightEditableContent: CharSequence
        get() = rightEditText.text

    init {
        LayoutInflater.from(context).inflate(R.layout.row_list_item, this, true)

        tv_title = findViewById(R.id.tv_title)
        iv_arrow = findViewById(R.id.iv_arrow)
        imageView = findViewById(R.id.tv_image)
        imageView = findViewById(R.id.tv_image)
        rightEditText = findViewById(R.id.et_edit)
        tv_right_desc = findViewById(R.id.tv_right_desc)
        checkBox = findViewById(R.id.checkBox)
    }

    fun setType(type: Int, clickable: Boolean = true) {
        tv_right_desc.visibility = View.GONE
        imageView!!.visibility = View.GONE
        rightEditText.visibility = View.GONE
        checkBox.visibility = View.GONE

        this.flag = type
        imageView!!.visibility = if (flag and FLAG_RIGHT_SHOW_IMAGE != 0) View.VISIBLE else View.GONE
        tv_right_desc.visibility = if (flag and FLAG_RIGHT_SHOW_TEXT != 0) View.VISIBLE else View.GONE
        if (flag and FLAG_RIGHT_EDITABLE != 0) {
            iv_arrow.visibility = View.GONE
            rightEditText.visibility = View.VISIBLE
        }

        if (flag and FLAG_RIGHT_CHECKABLE != 0) {
            iv_arrow.visibility = View.GONE
            checkBox.visibility = View.VISIBLE
        }

        if (!clickable) {
            iv_arrow.visibility = View.INVISIBLE
        }
    }

    fun setTitle(text: CharSequence) {
        tv_title.text = text
    }

    fun setRightDesc(text: String) {
        tv_right_desc.text = text
    }

    fun setRightEditableText(text: CharSequence) {
        rightEditText.setText(text)
    }

    fun setCheckedChangeListener(listener: CompoundButton.OnCheckedChangeListener?) {
        checkBox.setOnCheckedChangeListener(listener)
    }

    fun setChecked(flag: Boolean) {
        checkBox.isChecked = flag
    }

    companion object {
        val FLAG_NONE = 0      //仅仅左边的title
        val FLAG_RIGHT_SHOW_IMAGE = 1
        val FLAG_RIGHT_SHOW_TEXT = 1 shl 1
        val FLAG_RIGHT_EDITABLE = 1 shl 2
        val FLAG_RIGHT_CHECKABLE = 1 shl 3
    }
}
